### PR TITLE
RFC: ENH: make scipy.spatial.ConvexHull interruptible

### LINKED ISSUE #12219: spatial.ConvexHull is uninterruptible when running
It seems that when `spatial.ConvexHull` is running it cannot be interrupted with `SIGINT`.

#### Reproducing code example:

Feed the following array into `spatial.ConvexHull` and try to interrupt.

<details>
<summary>Example data</summary>

```python
D=array([[ 7.603686e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  7.603686e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  7.603686e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  7.603686e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  8.936956e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         8.936956e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  8.936956e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  8.936956e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  5.410814e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  5.410814e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         5.410814e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.410814e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00,  7.011844e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  7.011844e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  7.011844e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         7.011844e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  4.518964e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  4.518964e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  4.518964e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  4.518964e+00],
       [-3.662633e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.873305e-01,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.873305e-01,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  5.873305e-01,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00, -3.628802e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         5.819055e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.819055e-01,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.819055e-01,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00, -3.757803e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  6.025917e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         6.025917e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  6.025917e-01,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00, -3.682054e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  5.904448e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  5.904448e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         5.904448e-01,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00, -3.825832e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  6.135007e-01,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  6.135007e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  6.135007e-01],
       [ 5.873305e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00, -3.662633e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.873305e-01,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  5.873305e-01,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  5.819055e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
        -3.628802e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.819055e-01,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.819055e-01,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  6.025917e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00, -3.757803e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         6.025917e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  6.025917e-01,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00,  5.904448e-01,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00, -3.682054e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  5.904448e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         5.904448e-01,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  6.135007e-01,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00, -3.825832e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  6.135007e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  6.135007e-01],
       [ 5.873305e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.873305e-01,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00, -3.662633e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  5.873305e-01,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  5.819055e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         5.819055e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00, -3.628802e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.819055e-01,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  6.025917e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  6.025917e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
        -3.757803e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  6.025917e-01,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00,  5.904448e-01,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  5.904448e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00, -3.682054e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         5.904448e-01,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  6.135007e-01,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  6.135007e-01,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00, -3.825832e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  6.135007e-01],
       [ 5.873305e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.873305e-01,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.873305e-01,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00, -3.662633e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  5.819055e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         5.819055e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  5.819055e-01,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00, -3.628802e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  6.025917e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  6.025917e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         6.025917e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00, -3.757803e+00,
         0.000000e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00,  5.904448e-01,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  5.904448e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  5.904448e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
        -3.682054e+00,  0.000000e+00],
       [ 0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,  6.135007e-01,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  0.000000e+00,  6.135007e-01,  0.000000e+00,  0.000000e+00,
         0.000000e+00,  0.000000e+00,  6.135007e-01,  0.000000e+00,  0.000000e+00,  0.000000e+00,
         0.000000e+00, -3.825832e+00],
       [ 1.343122e+00, -6.418205e-02,  1.768226e+00,  1.064634e+00,  1.590279e+00,  1.343122e+00,
        -6.418205e-02,  1.768226e+00,  1.064634e+00,  1.590279e+00,  1.343122e+00, -6.418205e-02,
         1.768226e+00,  1.064634e+00,  1.590279e+00,  1.343122e+00, -6.418205e-02,  1.768226e+00,
         1.064634e+00,  1.590279e+00],
       [-8.810935e-01,  4.210370e-02, -1.159964e+00, -6.984042e-01, -1.043230e+00,  1.412899e-01,
        -6.751642e-03,  1.860088e-01,  1.119943e-01,  1.672897e-01,  1.412899e-01, -6.751642e-03,
         1.860088e-01,  1.119943e-01,  1.672897e-01,  1.412899e-01, -6.751642e-03,  1.860088e-01,
         1.119943e-01,  1.672897e-01],
       [ 1.412899e-01, -6.751642e-03,  1.860088e-01,  1.119943e-01,  1.672897e-01, -8.810935e-01,
         4.210370e-02, -1.159964e+00, -6.984042e-01, -1.043230e+00,  1.412899e-01, -6.751642e-03,
         1.860088e-01,  1.119943e-01,  1.672897e-01,  1.412899e-01, -6.751642e-03,  1.860088e-01,
         1.119943e-01,  1.672897e-01],
       [ 1.412899e-01, -6.751642e-03,  1.860088e-01,  1.119943e-01,  1.672897e-01,  1.412899e-01,
        -6.751642e-03,  1.860088e-01,  1.119943e-01,  1.672897e-01, -8.810935e-01,  4.210370e-02,
        -1.159964e+00, -6.984042e-01, -1.043230e+00,  1.412899e-01, -6.751642e-03,  1.860088e-01,
         1.119943e-01,  1.672897e-01],
       [ 1.412899e-01, -6.751642e-03,  1.860088e-01,  1.119943e-01,  1.672897e-01,  1.412899e-01,
        -6.751642e-03,  1.860088e-01,  1.119943e-01,  1.672897e-01,  1.412899e-01, -6.751642e-03,
         1.860088e-01,  1.119943e-01,  1.672897e-01, -8.810935e-01,  4.210370e-02, -1.159964e+00,
        -6.984042e-01, -1.043230e+00],
       [ 1.065574e-01,  1.082666e+00,  1.211024e+00,  7.435147e-01,  2.417595e+00,  1.065574e-01,
         1.082666e+00,  1.211024e+00,  7.435147e-01,  2.417595e+00,  1.065574e-01,  1.082666e+00,
         1.211024e+00,  7.435147e-01,  2.417595e+00,  1.065574e-01,  1.082666e+00,  1.211024e+00,
         7.435147e-01,  2.417595e+00],
       [-7.182555e-02, -7.297762e-01, -8.162966e-01, -5.011698e-01, -1.629592e+00,  1.151776e-02,
         1.170251e-01,  1.308992e-01,  8.036631e-02,  2.613172e-01,  1.151776e-02,  1.170251e-01,
         1.308992e-01,  8.036631e-02,  2.613172e-01,  1.151776e-02,  1.170251e-01,  1.308992e-01,
         8.036631e-02,  2.613172e-01],
       [ 1.151776e-02,  1.170251e-01,  1.308992e-01,  8.036631e-02,  2.613172e-01, -7.182555e-02,
        -7.297762e-01, -8.162966e-01, -5.011698e-01, -1.629592e+00,  1.151776e-02,  1.170251e-01,
         1.308992e-01,  8.036631e-02,  2.613172e-01,  1.151776e-02,  1.170251e-01,  1.308992e-01,
         8.036631e-02,  2.613172e-01],
       [ 1.151776e-02,  1.170251e-01,  1.308992e-01,  8.036631e-02,  2.613172e-01,  1.151776e-02,
         1.170251e-01,  1.308992e-01,  8.036631e-02,  2.613172e-01, -7.182555e-02, -7.297762e-01,
        -8.162966e-01, -5.011698e-01, -1.629592e+00,  1.151776e-02,  1.170251e-01,  1.308992e-01,
         8.036631e-02,  2.613172e-01],
       [ 1.151776e-02,  1.170251e-01,  1.308992e-01,  8.036631e-02,  2.613172e-01,  1.151776e-02,
         1.170251e-01,  1.308992e-01,  8.036631e-02,  2.613172e-01,  1.151776e-02,  1.170251e-01,
         1.308992e-01,  8.036631e-02,  2.613172e-01, -7.182555e-02, -7.297762e-01, -8.162966e-01,
        -5.011698e-01, -1.629592e+00]])
```
</details>

#### Error messages
```
% python test.py
^C^C^C^C^C^C^CTraceback (most recent call last):
  File "test.py", line 146, in <module>
    spatial.ConvexHull(D)
  File "qhull.pyx", line 2430, in scipy.spatial.qhull.ConvexHull.__init__
  File "qhull.pyx", line 1551, in scipy.spatial.qhull._QhullUser.__init__
  File "qhull.pyx", line 2463, in scipy.spatial.qhull.ConvexHull._update
  File "qhull.pyx", line 1580, in scipy.spatial.qhull._QhullUser._update
  File "/usr/lib/python3.8/site-packages/numpy/core/_methods.py", line 32, in _amin
    def _amin(a, axis=None, out=None, keepdims=False,
KeyboardInterrupt

python test.py  146.06s user 0.62s system 98% cpu 2:28.72 total
```

#### Scipy/Numpy/Python version information:
```
1.4.1 1.18.4 sys.version_info(major=3, minor=8, micro=3, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by hexchain**
#### Reference issue
Closes gh-12219.

#### What does this implement/fix?
Use `cysignals` module to allow interrupting C routines in `spatial.ConvexHull`.

#### Additional information
Not sure if a new dependency is what we want, but it works well in this case.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -3,6 +3,7 @@ requires = [
     "wheel",
     "setuptools",
     "Cython>=0.29.18",
+    "cysignals",
```
Comment (tylerjereddy) [inline pyproject.toml:6]: As you've alluded to in the PR description, addition of a new dependency to SciPy would indeed require some discussion.

@pv and @rgommers probably for initial thoughts on adding a dependency to improve ctrl+C handling in Cython.

I'm probably a little on the "against" side if we could have something that isn't too ugly internally instead, but let's let others chime in.

Comment (hexchain) [reply pyproject.toml:6]: I should have made it clearer that this is nothing more than an RFC in its current state. The sig_on sig_off indeed looks ugly.

**Comment by rgommers** [comment]
This would only be a build time dependency, and the package seems well-maintained and is part of SageMath. I'd be fine with adding it I think. We probably have similar Ctrl-C issues in other Cython code that's potentially taking a long time to run.

Disclaimer: I'm not familiar with the package.

**Comment by rgommers** [comment]
I asked for input on the mailing list: https://mail.python.org/pipermail/scipy-dev/2020-July/024283.html

**Comment by rkern** [comment]
`cysignals` is LGPLed, so I think this is a non-starter.

**Comment by hexchain** [comment]
I looked at their issue tracker, it seems that 

1. It currently does not support `-D_FORTIFY_SOURCE=2`, which is a common preprocessor flag used by multiple distributions for security reasons (It adds build-time and run-time checks to detect some memory safety problems).
2. The reason for 1 is that `cysignals` "abuses the stack in ways that violate some `-D_FORTIFY_SOURCE` checks" ([link](https://github.com/sagemath/cysignals/issues/120#issuecomment-613951472)).
3. It currently does not support multithreading.

Plus the license problem just mentioned above, it seems that `cysignals` may not fit for the purpose of enhancing signal processing in long-running processes.

I'm closing this PR then. But I would still like to see interruptible native routines in the future. Do you think the original issue can be changed to track this?

**Comment by rgommers** [comment]
Ah, somehow I managed to miss that licensing issue. Thanks @rkern.

> But I would still like to see interruptible native routines in the future, preferably in a less magical way. Do you think the original issue can be changed to track this?

yes that seems fine to me.

**Pull request closed without merging**